### PR TITLE
UBERF-8169: Account on postgres

### LINKED ISSUE #6622: Account service on postgres
<sub><a href="https://front.hc.engineering/guest/platform?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NmVhYTc1NjU1YmFlNTFmZjU5MjExMTMiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InBsYXRmb3JtIn0._KY9t2VZ8iGAC8lVQ87yTn4IjAGO9Lepbf7ppSc7vs0">Huly&reg;: <b>UBERF-8169</b></a></sub>

### PR DESCRIPTION
**Pull request by lexiv0re**
* Adds support for running account on postgres
* Streamlines usage of `MONGO_URL` and `DB_URL`. They always allow a single value now. `DB_URL` is the primary DB url required everywhere the db is needed while mongo is now an additional url for places where mongo is still used directly.

Note: I renamed `getWorkspaceDB` -> `getWorkspaceMongoDB` to have more clear picture of where we do still have direct Mongo uses.

Related to: https://front.hc.engineering/workbench/platform/tracker/UBERF-8169
Closes https://github.com/hcengineering/platform/issues/6622

<sub><a href="https://front.hc.engineering/guest/platform?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NmY1NThkNzFjYjUwNjYzZTU5MzMxMWMiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InBsYXRmb3JtIn0.1I_yPp0EgxsYDUur2pliEAY43ecFdtUXKVokdTFUpho">Huly&reg;: <b>UBERF-8293</b></a></sub>

### CODE REVIEW

File: dev/tool/src/index.ts [hunk 1]
```diff
@@ -540,7 +542,9 @@ export function devTool (
                     lastVisitDays
                   })
                   if (cmd.remove) {
-                    await dropWorkspaceFull(toolCtx, db, client, null, ws.workspace, adapter)
+                    const client = getMongoClient(mongodbUri)
+                    const _client = await client.getClient()
+                    await dropWorkspaceFull(toolCtx, db, _client, null, ws.workspace, adapter)
```
Comment (haiodo) [inline dev/tool/src/index.ts]: Client not closed

File: dev/tool/src/index.ts [hunk 2]
```diff
@@ -565,14 +569,16 @@ export function devTool (
...
           const ws = await getWorkspaceById(db, workspace)
           if (ws === null) {
             console.log('no workspace exists')
             return
           }
           if (cmd.full) {
-            await dropWorkspaceFull(toolCtx, db, client, null, workspace, storageAdapter)
+            const client = getMongoClient(mongodbUri)
+            const _client = await client.getClient()
+            await dropWorkspaceFull(toolCtx, db, _client, null, workspace, storageAdapter)
```
Comment (haiodo) [inline dev/tool/src/index.ts]: client not closed

File: dev/tool/src/index.ts [hunk 3]
```diff
@@ -587,10 +593,12 @@ export function devTool (
...
       const { mongodbUri } = prepareTools()
       await withStorage(mongodbUri, async (storageAdapter) => {
-        await withDatabase(mongodbUri, async (db, client) => {
+        await withDatabase(mongodbUri, async (db) => {
           for (const workspace of await listWorkspacesByAccount(db, email)) {
             if (cmd.full) {
-              await dropWorkspaceFull(toolCtx, db, client, null, workspace.workspace, storageAdapter)
+              const client = getMongoClient(mongodbUri)
+              const _client = await client.getClient()
+              await dropWorkspaceFull(toolCtx, db, _client, null, workspace.workspace, storageAdapter)
```
Comment (haiodo) [inline dev/tool/src/index.ts]: client not closed

File: dev/tool/src/index.ts [hunk 4]
```diff
@@ -618,12 +626,14 @@ export function devTool (
...
 
       await withStorage(mongodbUri, async (storageAdapter) => {
-        await withDatabase(mongodbUri, async (db, client) => {
+        await withDatabase(mongodbUri, async (db) => {
           const workspacesJSON = await listWorkspacesPure(db)
           for (const ws of workspacesJSON) {
             const lastVisit = Math.floor((Date.now() - ws.lastVisit) / 1000 / 3600 / 24)
             if (lastVisit > 30) {
-              await dropWorkspaceFull(toolCtx, db, client, null, ws.workspace, storageAdapter)
+              const client = getMongoClient(mongodbUri)
```
Comment (haiodo) [inline dev/tool/src/index.ts]: client not closed

File: dev/tool/src/index.ts [hunk 5]
```diff
@@ -1171,9 +1183,12 @@ export function devTool (
     .option('--all', 'Show all files', false)
     .action(async (cmd: { workspace: string, disabled: boolean, all: boolean }) => {
       const { mongodbUri } = prepareTools()
-      await withDatabase(mongodbUri, async (db, client) => {
+      await withDatabase(mongodbUri, async (db) => {
         await withStorage(mongodbUri, async (adapter) => {
           try {
+            const client = getMongoClient(mongodbUri)
+            const _client = await client.getClient()
```
Comment (haiodo) [inline dev/tool/src/index.ts]: client not closed

File: dev/tool/src/index.ts [hunk 6]
```diff
@@ -1532,7 +1547,9 @@ export function devTool (
 
             const { pipeline } = await getServerPipeline(toolCtx, txes, mongodbUri, dbUrl, wsUrl)
 
-            await migrateMarkup(toolCtx, adapter, wsId, client, pipeline, parseInt(cmd.concurrency))
+            const client = getMongoClient(mongodbUri)
```
Comment (haiodo) [inline dev/tool/src/index.ts]: client not closed

File: server/account/src/collections/postgres.ts [hunk 7]
```diff
@@ -0,0 +1,678 @@
...
+      )`
+    )
+
+    await this.client.query(`
+      CREATE INDEX ${this.name}_workspace ON ${this.name} ("workspace")
+    `)
+
+    // NOTE: as SKIP LOCKED is not supported in Cockroachdb, we need our own skip locked implementation
+    await this.client.query(`
+      CREATE OR REPLACE FUNCTION is_ws_locked(varchar(255)) RETURNS BOOLEAN AS $$
```
Comment (haiodo) [inline server/account/src/collections/postgres.ts]: Could we make it simplier???

Comment (lexiv0re) [reply server/account/src/collections/postgres.ts]: Unfortunately I didn't find a better way to implement effective queueing as SKIP LOCKED is not supported by Cockroach. Do you have any ideas how to simplify it?

File: dev/tool/src/index.ts (hunk 2 above)
Comment (lexiv0re) [reply dev/tool/src/index.ts]: Good catch! thanks.

Comment (lexiv0re) [reply dev/tool/src/index.ts]: All fixed

File: server/account/src/collections/postgres.ts (hunk 7 above)
Comment (lexiv0re) [reply server/account/src/collections/postgres.ts]: Done

**Review by haiodo — APPROVED** [review]

**Pull request merged**